text/2020-09-28-memory-trace.md

```diff
@@ -0,0 +1,131 @@
+}
+```
+
+A memory trace is tree that records how much memory its children and itself
+uses. It doesn't need to match any function stacktrace, instead it should have
+logically meaningful layout. For example, memory usage should be divided into
+several components under the root scope: TiDB EndPoint, Transaction, Raft,
+gRPC etc. TiDB EndPoint can divide its children by queries, while Raft can
+divide memory by store and apply. Name are defined as number for better
+performance. In practice, it can be mapped to enumerates instead.
+
+`size` should be the memory the category holds. It should *own* the memory.
+
+To record a trace, we need to define providers.
+
+```rust
+trait MemoryTraceProvider {
+    fn trace(&mut self, dump: &mut MemoryTrace);
+}
+```
+
+The trace method will fill the trace with stats itself and its children have.
+
+To trigger all providers, we need a trace manager.
+
+```rust
+pub struct MemoryTraceManager;
+
+impl MemoryTraceManager {
+    pub fn snapshot(&self) -> MemoryTrace;
+    pub fn register_provider(&mut self, provider: Box<dyn MemoryTraceProvider + Send + 'static>);
+}
+
+```
+
+`register_provider` will store the provider and `snapshot` will create a trace
+using all the information from registered providers.
+
+### Output
+
+A trace is a rust struct, when dumping the result, it needs to be transformed
+into some readable format. For best readability, it can be output as a
+framegraph using the real name. It can also support the existing format from
+[minitrace][1] to utilize more existing UI front end.
+
+[1]: https://github.com/pingcap-incubator/minitrace-rust
+
+### Implementation example
+
+#### Tracing memory used by peer fsm
+
+The stats can be stored in `StoreMeta`, and be updated after infrequent
+existing tick or a specific new tick. We define an additional struct as
+Provider:
+
+```rust
+pub struct RaftStoreMemoryTraceProvider {
+    meta: Arc<Mutex<StoreMeta>>,
+}
+
+impl MemoryTraceProvider for RaftStoreMemoryTraceProvider {
+    fn trace(&mut self, dump: &mut MemoryTrace) {
+        let info = {
+            let m = meta.lock().unwrap();
+            m.memory_stats.clone()
+        };
+        for (peer_id, stat) in info {
+            // record stats.
+        }
+    }
+}
+```
+
+#### Tracing memory used by grpc
+
+gRPC already records its memory usage in quota, all we need to do is querying
+the quota.
+
+```rust
+pub struct GRpcMemoryTraceProvider {
+    quota: MemoryQuota,
+}
+
+impl MemoryTraceProvider for GRpcMemoryTraceProvider {
+    fn trace(&mut self, dump: &mut MemoryTrace) {
+        dump.set_size(quota.used());
+    }
+}
+```
+
+## Drawbacks
+
+Instrumentation can make code complicated.
+
+## Alternatives
+
+We can still keep using the profile tool we used to do and delivering profile
+build by default. It can have uncontrolled performance impact and doesn't have
+a clear view on historical allocations.
+
+## Unresolved questions
+
+How to ensure all memory allocations are traced?
```

Comment:
Can we have some more generic memory tracing, instead of manually implement one? I'm worried that a manually implemented one may *miss* actual memory leaks since these memory leaks are usually not noticed.

Reply:
We already have it in TiKV. According to the experiences, it's hard to used and requires recompilation and restart (and extra environment variables).

Even all the limitations are solved without hurting performance hopefully, it's still limited in functionality. For example, it can also miss allocations as stats are sample. It can't provide precise memory analyze by components like what RFC does. It also can't be used in QoS.

However, it's still useful when some places are missed by MemoryTraceProvider. It can be used as the last resort for memory leaks.

Reply:
Is the reason that it is hard to use because such a a mechanism doesn't properly annotate allocations? Could we annotate allocations in our code (for example wrap Box::new) and would that cover a lot of annotations or would we have many allocations from libraries?

Reply:
What does "annotate" mean? `Box::new` will call `je_malloc` under the hook and be recorded into profile if sampled. It's hard to use because of recompilation and restart. Enabling profile for a long time may not be practical for online system.

Reply:
I am wondering if we should replace usages of Box::new with our own wrapper that helps to record memory usage. This would help to ensure we don't miss allocations. But maybe there is no way to make this an improvement over the existing je_malloc handle?

Reply:
Yes. In my opinion, we care what holds the memory rather than allocating, which can't be solved by tagging allocations.

Reply:
But if we control tagging allocations we could tag them with information about who holds them?

Reply:
For example, memory allocated in grpc can go through all the way grpc, transaction and raftstore. Ownership is unpredictable.